strawberryfields/engine.py

```diff
@@ -277,15 +272,57 @@ def _broadcast_nones(val, dim):
                 p = p.compile(target, **compile_options)
             p.lock()
 
-            self._run_program(p, **kwargs)
-            shots = kwargs.get("shots", 1)
-            self.samples = [_broadcast_nones(p.reg_refs[k].val, shots) for k in sorted(p.reg_refs)]
+            _, values = self._run_program(p, **kwargs)
             self.run_progs.append(p)
 
+            # check number of measurments and combine samples if needed
+            if len(values) > 1:
+                self._combine_samples(values, p)
+            elif len(values) == 1:
+                self.samples = values[0]
+            else:  # if no measurement was made
+                self.samples = np.array([[]])
```

Comment:
Just out of curiosity, should we be returning `None` here? By returning an empty NumPy array, we are returning something with shape `(1, 0)`, which violates the `(shots, modes)` convention.

On the other hand, `bool(np.array([[]])) == False`, so it would work well in `if` statements (`if results.samples:`). However, this raises a deprecation warning, so this behavior will likely be dropped soon 🤔

Reply:
having a `None` value here sounds reasonable, the if statement could then look as `if results.samples is not None:`

Reply:
I don't know what would be the best option here. Just changing to `self.samples=None` in the engine would cause issues when running sequential programs since it loops over `self.samples`. This could either be fixed by changing how the results are saved in `results.py`; or by simply not returning anything from the engine, and thus "returning" None; or by adding an `if self.samples is not None:` to line 263 in `engine.py` (above the `for k, v in enumerate(self.samples):`).

> By returning an empty NumPy array, we are returning something with shape (1, 0), which violates the (shots, modes) convention.

An option would be to have `self.samples = np.empty((0, 0))`, which would actually follow the `(shots, measured_modes)` convention. :laughing:

Reply:
That's a fair point, haven't thought of it needing to be iterable! I've found [this comment](https://stackoverflow.com/a/14640048) useful and overall agree with what is described there, that is it's good to return the same datatype. With these points in mind, I think that your last point on having `self.samples = np.empty((0, 0))` does seem pretty reasonable and nice actually! :smile: